### PR TITLE
qa/workunits/ceph-helpers: display rejected string

### PR DESCRIPTION
**Pull request by tchaikov**
Signed-off-by: Kefu Chai <kchai@redhat.com>

### CODE REVIEW

**Comment by tchaikov** [comment]
```
/home/jenkins-build/build/workspace/ceph-pull-requests/qa/workunits/ceph-helpers.sh:1454: erasure_code_plugin_exists:  ceph osd erasure-code-profile set TESTPROFILE plugin=FAKE
Error EIO: load dlopen(/home/jenkins-build/build/workspace/ceph-pull-requests/build/lib/libec_FAKE.so): /home/jenkins-build/build/workspace/ceph-pull-requests/build/lib/libec_FAKE.so: cannot open shared object file: No such file or directory
/home/jenkins-build/build/workspace/ceph-pull-requests/qa/workunits/ceph-helpers.sh:1455: erasure_code_plugin_exists:  status=1
```

see https://jenkins.ceph.com/job/ceph-pull-requests/21899/consoleFull#-163083258740526d21-3511-427d-909c-dd086c0d1034

**Comment by dachary** [comment]
@tchaikov the idea is to not run another command to display the result because it may show something different, right ?

**Comment by tchaikov** [comment]
@dachary exactly.

**Comment by tchaikov** [comment]
@dachary @badone just updated the PR. the change was buggy.

**Comment by tchaikov** [comment]
and a side effect of this change is that we don't need to take the "pipefail option" into consideration when looking at this test anymore.

File: qa/workunits/ceph-helpers.sh [hunk 1]
```diff
@@ -1448,14 +1448,13 @@ function erasure_code_plugin_exists() {
...
-        # display why the string was rejected.
-        ceph osd erasure-code-profile set TESTPROFILE plugin=$plugin
-        status=1
-    else
-        status=0
+    local s=$(ceph osd erasure-code-profile set TESTPROFILE plugin=$plugin 2>&1)
+    local status=$?
+    if [ $status -eq 0 ]; then
         ceph osd erasure-code-profile rm TESTPROFILE
+    elif [ ! echo $s | grep --quiet "$grepstr" ]; then
```
Comment (dachary) [inline qa/workunits/ceph-helpers.sh]: <pre>
elif  ! echo $s | grep --quiet "$grepstr" ; then
   status=1
</pre>

**Comment by tchaikov** [comment]
```
/home/jenkins-build/build/workspace/ceph-pull-requests/qa/workunits/ceph-helpers.sh:1451: erasure_code_plugin_exists:  local 's=Error EIO: load dlopen(/home/jenkins-build/build/workspace/ceph-pull-requests/build/lib/libec_FAKE.so): /home/jenkins-build/build/workspace/ceph-pull-requests/build/lib/libec_FAKE.so: cannot open shared object file: No such file or directory'
/home/jenkins-build/build/workspace/ceph-pull-requests/qa/workunits/ceph-helpers.sh:1452: erasure_code_plugin_exists:  local status=0
/home/jenkins-build/build/workspace/ceph-pull-requests/qa/workunits/ceph-helpers.sh:1453: erasure_code_plugin_exists:  '[' 0 -eq 0 ']'
/home/jenkins-build/build/workspace/ceph-pull-requests/qa/workunits/ceph-helpers.sh:1454: erasure_code_plugin_exists:  ceph osd erasure-code-profile rm TESTPROFILE
erasure-code-profile TESTPROFILE does not exist
/home/jenkins-build/build/workspace/ceph-pull-requests/qa/workunits/ceph-helpers.sh:1459: erasure_code_plugin_exists:  return 0
/home/jenkins-build/build/workspace/ceph-pull-requests/qa/workunits/ceph-helpers.sh:1469: test_erasure_code_plugin_exists:  return 1
```

https://jenkins.ceph.com/job/ceph-pull-requests/21911/consoleFull#-209497759540526d21-3511-427d-909c-dd086c0d1034

so ceph failed to set the FAKE ec plugin, but it returned 0 nevertheless.

File: qa/workunits/ceph-helpers.sh [hunk 2]
```diff
@@ -1443,19 +1443,20 @@ function erasure_code_plugin_exists() {
     local plugin=$1
     local status
     local grepstr
+    local s
```
Comment (tchaikov) [inline qa/workunits/ceph-helpers.sh]: @badone @dachary i realized that `local` is a command by itself. so i should run it before assigning the output to `s`, now the captured return code should be correct.

**Comment by tchaikov** [comment]
@dachary @badone fixed and repushed.

**Comment by tchaikov** [comment]
```
/home/jenkins-build/build/workspace/ceph-pull-requests/qa/workunits/ceph-helpers.sh:1166: test_is_clean:  is_clean
//home/jenkins-build/build/workspace/ceph-pull-requests/qa/workunits/ceph-helpers.sh:1154: is_clean:  get_num_pgs
//home/jenkins-build/build/workspace/ceph-pull-requests/qa/workunits/ceph-helpers.sh:1097: get_num_pgs:  ceph --format xml status
//home/jenkins-build/build/workspace/ceph-pull-requests/qa/workunits/ceph-helpers.sh:1098: get_num_pgs:  xmlstarlet sel -t -m //pgmap/num_pgs -v .
/home/jenkins-build/build/workspace/ceph-pull-requests/qa/workunits/ceph-helpers.sh:1154: is_clean:  num_pgs=4
/home/jenkins-build/build/workspace/ceph-pull-requests/qa/workunits/ceph-helpers.sh:1155: is_clean:  test 4 '!=' 0
//home/jenkins-build/build/workspace/ceph-pull-requests/qa/workunits/ceph-helpers.sh:1156: is_clean:  get_num_active_clean
//home/jenkins-build/build/workspace/ceph-pull-requests/qa/workunits/ceph-helpers.sh:1061: get_num_active_clean:  local 'expression=('
//home/jenkins-build/build/workspace/ceph-pull-requests/qa/workunits/ceph-helpers.sh:1062: get_num_active_clean:  expression+='contains(.,'\''active'\'') and '
//home/jenkins-build/build/workspace/ceph-pull-requests/qa/workunits/ceph-helpers.sh:1063: get_num_active_clean:  expression+='contains(.,'\''clean'\'') and '
//home/jenkins-build/build/workspace/ceph-pull-requests/qa/workunits/ceph-helpers.sh:1064: get_num_active_clean:  expression+='not(contains(.,'\''stale'\''))'
//home/jenkins-build/build/workspace/ceph-pull-requests/qa/workunits/ceph-helpers.sh:1065: get_num_active_clean:  expression+=')'
//home/jenkins-build/build/workspace/ceph-pull-requests/qa/workunits/ceph-helpers.sh:1069: get_num_active_clean:  ceph --format xml pg dump pgs
//home/jenkins-build/build/workspace/ceph-pull-requests/qa/workunits/ceph-helpers.sh:1070: get_num_active_clean:  xmlstarlet sel -t -m '//pg_stat/state[(contains(.,'\''active'\'') and contains(.,'\''clean'\'') and not(contains(.,'\''stale'\'')))]' -v . -n
//home/jenkins-build/build/workspace/ceph-pull-requests/qa/workunits/ceph-helpers.sh:1071: get_num_active_clean:  grep -cv '^$'
/home/jenkins-build/build/workspace/ceph-pull-requests/qa/workunits/ceph-helpers.sh:1156: is_clean:  test 4 = 4
/home/jenkins-build/build/workspace/ceph-pull-requests/qa/workunits/ceph-helpers.sh:1166: test_is_clean:  return 1
/home/jenkins-build/build/workspace/ceph-pull-requests/qa/workunits/ceph-helpers.sh:1649: run_tests:  return 1
```
see https://jenkins.ceph.com/job/ceph-pull-requests/21917/console

**Comment by tchaikov** [comment]
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/osd/osd-scrub-repair.sh:66: add_something:  rados --pool ecpool put SOMETHING td/osd-scrub-repair/ORIGINAL
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/osd/osd-scrub-repair.sh: line 49: 20401 Terminated              rados --pool $poolname put $obj $dir/ORIGINAL
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/osd/osd-scrub-repair.sh:66: add_something:  return 1
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/osd/osd-scrub-repair.sh:163: corrupt_and_repair_erasure_coded:  return 1
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/osd/osd-scrub-repair.sh:241: TEST_corrupt_and_repair_jerasure:  return 1
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/osd/osd-scrub-repair.sh:45: run:  return 1
```

retest this please.

**Comment by tchaikov** [comment]
retest this please.

File: qa/workunits/ceph-helpers.sh [hunk 3]
```diff
@@ -1163,7 +1163,6 @@ function test_is_clean() {
     run_mon $dir a --osd_pool_default_size=1 || return 1
     run_mgr $dir x || return 1
     run_osd $dir 0 || return 1
-    ! is_clean || return 1
```
Comment (dachary) [inline qa/workunits/ceph-helpers.sh]: good catch !

**Comment by smithfarm** [comment]
@tchaikov Guess this needs to be backported to kraken? See #14523 which is waiting for it?

**Comment by tchaikov** [comment]
@smithfarm yes, i think so.

**Comment by smithfarm** [comment]
@kefu @badone Re-running make check at #14523 - if your theory is right, make check should succeed there.

**Pull request merged**